Userland/Libraries/LibWeb/Painting/VideoPaintable.cpp

```diff
@@ -74,19 +74,104 @@ void VideoPaintable::paint(PaintContext& context, PaintPhase phase) const
+        // The video element represents its poster frame, if any, or else the first frame of the video.
+        representation = poster_frame ? Representation::PosterFrame : Representation::FirstVideoFrame;
+    }
+
+    // -> When the video element is paused, and the frame of video corresponding to the current playback position
+    //    is not available (e.g. because the video is seeking or buffering)
+    // -> When the video element is neither potentially playing nor paused (e.g. when seeking or stalled)
+    else if (
+        (video_element.paused() && current_playback_position != current_frame.position)
+        || (!video_element.potentially_playing() && !video_element.paused())) {
+        // The video element represents the last frame of the video to have been rendered.
+        representation = Representation::LastRenderedVideoFrame;
+    }
+
+    // -> When the video element is paused
+    else if (video_element.paused()) {
+        // The video element represents the frame of video corresponding to the current playback position.
+        representation = Representation::CurrentVideoFrame;
+    }
+
+    // -> Otherwise (the video element has a video channel and is potentially playing)
+    else {
+        // The video element represents the frame of video at the continuously increasing "current" position. When the
+        // current playback position changes such that the last frame rendered is no longer the frame corresponding to
+        // the current playback position in the video, the new frame must be rendered.
+        representation = Representation::CurrentVideoFrame;
+    }
 
-    if (auto const& bitmap = layout_box().dom_node().current_frame().frame) {
-        context.painter().draw_scaled_bitmap(video_rect.to_type<int>(), *bitmap, bitmap->rect(), 1.0f, to_gfx_scaling_mode(computed_values().image_rendering()));
+    auto paint_frame = [&](auto const& frame) {
+        context.painter().draw_scaled_bitmap(video_rect.to_type<int>(), *frame, frame->rect(), 1.0f, to_gfx_scaling_mode(computed_values().image_rendering()));
+    };
 
+    auto paint_transparent_black = [&]() {
+        static constexpr auto transparent_black = Gfx::Color::from_argb(0x00'00'00'00);
+        context.painter().fill_rect(video_rect.to_type<int>(), transparent_black);
+    };
+
+    auto paint_user_agent_controls = video_element.has_attribute(HTML::AttributeNames::controls) || video_element.is_scripting_disabled();
+
+    switch (representation) {
+    case Representation::FirstVideoFrame:
+    case Representation::CurrentVideoFrame:
+    case Representation::LastRenderedVideoFrame:
+        // FIXME: We likely need to cache all (or a subset of) decoded video frames along with their position. We at least
+        //        will need the first video frame and the last-rendered video frame.
+        if (current_frame.frame)
+            paint_frame(current_frame.frame);
```

Comment:
I'm skeptical of whether we actually need to cache any frames. There shouldn't be any distinction between the `current_frame` image and the last frame rendered.

Also, caching the first frame _shouldn't_ be necessary, since the playback manager should in all cases decode and send for presentation the first frame when it is paused on timestamp 0. If it doesn't, that's a bug we definitely should be aware of. However, I'm not super opposed to caching that still, since in very high resolution video, it may take much more than the 16ms target to decode the frame, but conversely that also means it increases the memory consumption quite a lot.

Reply:
Right, I'm not 100% sure how much (if any) needs to be cached, and we're basically 1-to-1 between the current time and current frame right now anyways, which is why I didn't add any caching here.

I think this will come more into play when we don't have the entire video file in memory - i.e. the spec mentions we are allowed to hold onto a subset of the file and discard old data (but doesn't recommend *how* to do that of course). But since the fetching/decoding part of the spec is generalized like that, the painting is generalized to match.

It's probably all a "we'll see what happens when we cross that bridge" scenario.